### PR TITLE
musl: 1.2.3 -> 1.2.5

### PR DESCRIPTION
**Pull request by alyssais**
From a packaging perspective, this is the most important thing about this upgrade:

> Users who build applications from source against musl should be aware: 1.2.4 deprecates the "LFS64" interfaces, which were provided via macros (causing lots of frustration by breaking things; thus the motivation to get rid of them).
>
> Builds which break can be fixed for now by adding -D_LARGEFILE64_SOURCE to CFLAGS, but this too will be dropped in a later version.

[Source](https://fosstodon.org/@musl/110297136732196345)

I think we should see which packages are affected, and give it some reasonable amount of time for them to be fixed, because I don't think it's going to always be as simple as modifying CFLAGS — what do we do for rustc, for example?

We also have some work ahead of us to even identify broken packages.  I've done some test builds of the Spectrum live image, which heavily uses musl, to identify a few regressions, but since we don't have Hydra for musl, we're going to need people who use musl in Nixpkgs to do their own test builds to check whether any of the packages they care about are affected.  If you do so, add any more broken packages to the list in this post (or comment if you lack permissions).

- [x] LLVM #229442
- [x] libselinux https://github.com/NixOS/nixpkgs/pull/287471
- [x] acl #229440
- [x] xorg.libpciaccess #229441
- [x] rustc (and cross-compiled programs that use non-musl build rustc) #259989
- [x] gperftools https://github.com/NixOS/nixpkgs/pull/253556
- [x] gdb #229673
- [x] doxygen https://github.com/NixOS/nixpkgs/pull/292630
- [x] libredirect https://github.com/NixOS/nixpkgs/pull/292616
- [x] numactl https://github.com/NixOS/nixpkgs/pull/281724
- [x] autogen https://github.com/NixOS/nixpkgs/pull/292617
- [x] libeatmydata https://github.com/stewartsmith/libeatmydata/pull/36 https://github.com/NixOS/nixpkgs/pull/324772
- [x] bpftools
- [x] libbpf
- [x] gst_all_1.gstreamer
- [x] meson https://github.com/NixOS/nixpkgs/pull/298047
- [x] elfutils https://github.com/NixOS/nixpkgs/pull/293517
- [x] yara https://github.com/NixOS/nixpkgs/pull/306531
- [x] directfb https://github.com/NixOS/nixpkgs/pull/309583
- [x] systemd #311995
- [x] gpgme ([Alpine patch](https://git.alpinelinux.org/aports/plain/community/gpgme/lfs64.patch?id=13ae3e5ff505861073a345e579a3362214265581)) ([Proposed upstream patch](https://dev.gnupg.org/D600)) https://github.com/NixOS/nixpkgs/pull/324771
- [x] mblaze https://github.com/NixOS/nixpkgs/pull/306517
- [x] f2fs-tools https://github.com/NixOS/nixpkgs/pull/306514
- [x] ocamlPackages.uring https://github.com/NixOS/nixpkgs/pull/307693
- [x] wgautomesh https://github.com/NixOS/nixpkgs/pull/321524
- [x] libfaketime https://github.com/NixOS/nixpkgs/pull/313663
- [x] aemu https://android-review.googlesource.com/c/platform/hardware/google/aemu/+/3105640/1, https://cyberchaos.dev/yuka/nixpkgs/-/commit/5e94b5a82ecbfcefd60c5f8f21d98ba7361eeac2 https://github.com/NixOS/nixpkgs/pull/330609
- [x] e2fsprogs ([Alpine patch](https://git.alpinelinux.org/aports/tree/main/e2fsprogs/lfs64.patch)) ([Upstream patch](https://github.com/tytso/e2fsprogs/commit/24181bb25e52fdecadbc7e434834e6ece85ff932)) https://github.com/NixOS/nixpkgs/pull/313657
- [x] ~fuse2  (issue in fuse headers manifests in e2fsprogs after updating) ([Void patch](https://github.com/void-linux/void-packages/blob/master/srcpkgs/fuse/patches/fuse-headers.patch))~ no other package seems to have that problem, and e2fsprogs is fixed now.
- [x] mdadm ([Alpine patch](https://git.alpinelinux.org/aports/tree/main/mdadm/lfs64.patch)) ([Proposed upstream patch](https://lore.kernel.org/linux-raid/20240220165158.3521874-1-raj.khem@gmail.com/)) https://github.com/NixOS/nixpkgs/pull/313936
- [x] xfsprogs https://termbin.com/8vdm #304745  
- [x] multipath-tools https://github.com/NixOS/nixpkgs/pull/313682
- [x] kmod https://github.com/NixOS/nixpkgs/pull/322720
- [x] mumble https://github.com/NixOS/nixpkgs/pull/322756
- [ ] anything that uses protobuf 21, (wontfix, mark as broken: https://cyberchaos.dev/yuka/nixpkgs/-/commit/37d0ec2c0a91314de3177ab1a9ce838693b020a0) https://github.com/NixOS/nixpkgs/pull/339631
- [x] nfs-utils (https://github.com/NixOS/nixpkgs/pull/278160, https://cyberchaos.dev/yuka/nixpkgs/-/commit/c53c94dbe882936eeb2956d6f5655540cc1debe4)
- [x] openjdk17 (https://github.com/NixOS/nixpkgs/pull/327135, https://cyberchaos.dev/yuka/nixpkgs/-/commit/0f58ffa8cf636cece3079b68eebf2c5082a50f01)
- [x] gn https://github.com/NixOS/nixpkgs/pull/339794
- [x] intel-gpu-tools https://github.com/NixOS/nixpkgs/pull/341576
- [x] catfs https://github.com/NixOS/nixpkgs/pull/325088

### New! with GCC 14

- [x] attr https://github.com/NixOS/nixpkgs/pull/342008
- [x] bluez https://github.com/NixOS/nixpkgs/pull/342281
- [x] elogind https://github.com/NixOS/nixpkgs/pull/342236
- [x] kexec-tools https://github.com/NixOS/nixpkgs/pull/342004
- [x] libapparmor https://github.com/NixOS/nixpkgs/pull/342013
- [x] util-linux https://github.com/NixOS/nixpkgs/pull/342019
- [x] iproute2 https://github.com/NixOS/nixpkgs/pull/342002

Where we can fix things without breaking things for musl 1.2.3, let's apply those fixes in separate PRs, so the work can be more easily spread out, and avoid this one becoming intimidatingly huge.

## Resources

- [Alpine Linux's fixes](https://git.alpinelinux.org/aports/log/?qt=grep&q=lfs64)

### CODE REVIEW

**Comment by alyssais** [comment]
https://github.com/NixOS/nixpkgs/pull/259989 includes the necessary Rust fixes, so will unblock this when merged.

File: pkgs/development/libraries/libtasn1/default.nix [hunk 1]
```diff
@@ -20,14 +20,6 @@ stdenv.mkDerivation rec {
     sha256 = "sha256-FhPwrBz0hNbsDOO4wG1WJjzHJC8cI7MNgtI940WmP3o=";
   };
 
-  # Patch borrowed from alpine to work around a specific test failure with musl libc
-  # Upstream is patching this test in their own CI because that CI is using alpine and thus musl
-  # https://github.com/gnutls/libtasn1/commit/06e7433c4e587e2ba6df521264138585a63d07c7#diff-037ea159eb0a7cb0ac23b851e66bee30fb838ee8d0d99fa331a1ba65283d37f7R293
-  patches = lib.optional stdenv.hostPlatform.isMusl (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libtasn1/default.nix:26]: We can drop fetchpatch in line 4

**Comment by alyssais** [comment]
1.2.5 is out now, but it doesn't look like anything in that update will be nearly as hard as LFS in 1.2.4, so I'm now aiming to go straight to 1.2.5 with this.  Help with reviewing the package fixes (the libselinux one has been open for a while) would be very helpful with getting this moving.

**Comment by fzakaria** [comment]
I opened https://github.com/NixOS/nixpkgs/pull/313439 didn't see it -- will close.

**Comment by fzakaria** [comment]
is there a Nix cache for musl ?
I'm stuck always rebuilding the world :)

**Comment by fzakaria** [comment]
Would be cool to cook up an update script also for this.

**Comment by fzakaria** [comment]
Let me know how I can help.

I ran into issues building open office for musl due to NULL being #define as nullptr from [this patch series](https://www.openwall.com/lists/musl/2021/12/24/1) -- figured I'd see if we were trying to upgrade first.

Let me know how I can help.

**Comment by alyssais** [comment]
> Let me know how I can help.

Find packages that currently build, but don't build with this PR applied.  (Or if you can't find any, let me know, that's also good data.)

**Comment by fzakaria** [comment]
Does nixpkgs with musl (non-static) even build at 1.2.3 ?
I found a few projects that were broken.

Is there a separate musl room on matrix or discord?

**Comment by fzakaria** [comment]
@alyssais I'll start a running list in this thread:

### Trying to build openmpi

* pkgs/os-specific/linux/libpsm2/default.nix does not build.
   It uses `drand48_data` which is a non-standard libc extension in GLIBC.
  ```
  ptl_ips/ips_proto_recv.c:895:37: error: storage size of ‘drand48_data’ isn’t known
  895 |                 struct drand48_data drand48_data;
  ```

### Trying to build libreoffice

* pkgs/development/libraries/CoinMP/default.nix
   It uses a reinterpreter_cast that causes failures as it's non compliant and musl disallows it.

**Comment by alyssais** [comment]
This isn't the place to collect a list of all packages that fail to build with musl.  We're specifically interested in packages that _do_ build with musl on current Nixpkgs, but fail with this PR applied.  What you have posted look like packages that are already failing with musl, which are irrelevant for the purpose of upgrading.

**Comment by alyssais** [comment]
@yu-re-ka can we keep them all in the big list in the PR body?  Otherwise they're going to get hard to keep track of when GitHub starts hiding comments.

**Comment by AndersonTorres** [comment]
There is a PR updating GN here:

https://github.com/NixOS/nixpkgs/pull/221306

**Comment by alyssais** [comment]
Unless there are any last minute objections, I intend to merge this with the next staging cycle. (There aren't enough rebuilds to force this into staging, but there are important fixes that are only in staging.)  Every package previously identified as regressing on update builds with musl 1.2.5 on staging, with the exception of nfs-utils, which doesn't work on master anyway and hasn't for a while.  Merging this is the best way to fix building pulseaudio with musl and GCC 14, and we're also increasingly seeing packages relying on features introduced in later versions of musl — there are two reverts for hacks we've had to do for 1.2.3 in this PR, and I also recently discovered that we need a newer musl for bcachefs-tools.

Since we don't have Hydra for musl, there will absolutely be a long tail of broken packages that we can't identify, but it's been a year and a half.  This is as good as it's going to get, and we can't hold back forever.

**Review by SuperSandro2000 — APPROVED** [review]
go for it 🎉

**Pull request merged**